drivers/usb/udc/udc_numaker.c

```diff
@@ -567,66 +814,248 @@ static void numaker_usbd_ep_th(const struct device *dev, uint32_t ep_hw_idx)
+	dmabuf_addr = numaker_usbd_buf_base(dev) + (base->STBUFSEG & USBD_STBUFSEG_STBUFSEG_Msk);
 	bytecpy(usrbuf, (uint8_t *)dmabuf_addr, 8ul);
+#elif DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_hsusbd)
+	*usrbuf = (uint8_t)(base->SETUP1_0 & 0xfful);
+	*(usrbuf + 1) = (uint8_t)((base->SETUP1_0 >> 8) & 0xfful);
+	*(usrbuf + 2) = (uint8_t)(base->SETUP3_2 & 0xfful);
+	*(usrbuf + 3) = (uint8_t)((base->SETUP3_2 >> 8) & 0xfful);
+	*(usrbuf + 4) = (uint8_t)(base->SETUP5_4 & 0xfful);
+	*(usrbuf + 5) = (uint8_t)((base->SETUP5_4 >> 8) & 0xfful);
+	*(usrbuf + 6) = (uint8_t)(base->SETUP7_6 & 0xfful);
+	*(usrbuf + 7) = (uint8_t)((base->SETUP7_6 >> 8) & 0xfful);
+#endif
 }
 
+#if DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_hsusbd)
+#if defined(CONFIG_UDC_NUMAKER_DMA)
+/* Transfer data between user buffer and USB buffer by DMA
+ *
+ * size_p holds size to copy/copied on input/output
+ */
+static int numaker_xusbd_ep_xfer_user_dma(struct numaker_xusbd_ep *ep_cur, uint8_t *usrbuf,
+					  uint32_t *size_p)
+{
+	const struct device *dev = ep_cur->dev;
+	const struct udc_numaker_config *config = dev->config;
+	struct udc_numaker_data *priv = udc_get_private(dev);
+	NUMAKER_XUSBD_TYPE *base = config->base;
+	int err;
+
+	/* Reset DMA semaphore */
+	k_sem_reset(&priv->sem_dma_done);
+
+	/* Reset DMA */
+	base->DMACNT = 0;
+	base->DMACTL = HSUSBD_DMACTL_DMARST_Msk;
+	base->DMACTL = 0;
+	base->BUSINTSTS = HSUSBD_BUSINTSTS_DMADONEIF_Msk;
+
+	/* DMA memory address */
+	base->DMAADDR = (uint32_t)usrbuf;
+
+	/* DMA transfer size */
+	base->DMACNT = *size_p;
+
+	/* DMA EP address */
+	base->DMACTL = USB_EP_DIR_IS_IN(ep_cur->addr)
+			       ? (HSUSBD_DMACTL_SVINEP_Msk | HSUSBD_DMACTL_DMARD_Msk)
+			       : 0;
+	base->DMACTL |= USB_EP_GET_IDX(ep_cur->addr) << HSUSBD_DMACTL_EPNUM_Pos;
+
+	/* Start DMA */
+	base->DMACTL |= HSUSBD_DMACTL_DMAEN_Msk;
+
+	/* Wait for DMA done */
+	err = k_sem_take(&priv->sem_dma_done, K_MSEC(CONFIG_UDC_NUMAKER_DMA_TIMEOUT_MS));
+	if (err != 0) {
+		err = -EIO;
+
+		/* Abort DMA for safe */
+		base->DMACNT = 0;
+		base->DMACTL = HSUSBD_DMACTL_DMARST_Msk;
+		base->DMACTL = 0;
+	}
+
+	return err;
+}
+#endif
+#endif
+
 /* Copy data to user buffer
  *
  * size_p holds size to copy/copied on input/output
  */
-static void numaker_usbd_ep_copy_to_user(struct numaker_usbd_ep *ep_cur, uint8_t *usrbuf,
+static int numaker_xusbd_ep_copy_to_user(struct numaker_xusbd_ep *ep_cur, uint8_t *usrbuf,
 					 uint32_t *size_p, uint32_t *rmn_p)
 {
 	const struct device *dev = ep_cur->dev;
-	USBD_EP_T *ep_base = numaker_usbd_ep_base(dev, ep_cur->ep_hw_idx);
+	const struct udc_numaker_config *config = dev->config;
+	NUMAKER_XUSBD_TYPE *base = config->base;
+	NUMAKER_XUSBD_EP_TYPE *ep_base = numaker_xusbd_ep_base(dev, ep_cur->ep_hw_idx);
 	uint32_t dmabuf_addr;
 	uint32_t data_rmn;
 
+	ARG_UNUSED(base);
+	ARG_UNUSED(dmabuf_addr);
+
 	__ASSERT_NO_MSG(size_p);
 	__ASSERT_NO_MSG(ep_cur->dmabuf_valid);
 
+#if DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_usbd)
```

Comment:
Something like `DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_usbd)` should not be used in the code, the driver has to support multiple instances of the controller on the SoC.